### ISSUE TITLE
openapi.json (swagger 2.0) definition file not parsed correctly in zap-api-scan.py

### ISSUE THREAD

**Issue by bravoman** [zaproxy/zaproxy#4072]
My **`openapi.json`** definition file looks like this:

```json
{
    "swagger": "2.0",
    "info": {
        "title": "BSS-API",
        "description": "Client BSS API",
        "version": "1.0.0"
    },
    "host": "api.client.tst",
    "basePath": "/",
    "schemes": [
        "https"
    ],
    "consumes": [
        "application/json",
        "application/javascript"
    ],
    "produces": [
        "application/json",
        "application/javascript"
    ],
    "paths": {
        "/availability/username_password_pair/{username}/{password}": {
            "get": {
                "summary": "Returns username password pair availability information",
                "responses": {
                    "200": {
                        "description": "Availability information"
                    },
                    "default": {
                        "description": "an \"unexpected\" error"
                    }
                }
            },
            "parameters": [
                {
                    "$ref": "#/securityDefinitions/api_key"
                },
                {
                    "name": "username",
                    "in": "path",
                    "required": true,
                    "type": "string"
                },
                {
                    "name": "password",
                    "in": "path",
                    "required": true,
                    "type": "string"
                }
            ]
        }
    },
    "definitions": {},
    "securityDefinitions": {
        "api_key": {
            "type": "apiKey",
            "name": "HTTP-apikey",
            "in": "header"
        }
    }
}
```

It gives me this inside the output when executing `zap-api-scan.py`: 

```

// ... some other non-related output

2017-11-22 16:22:01,280 Import OpenAPI File openapi.json
2017-11-22 16:22:01,707 Failed to import any URLs
2017-11-22 16:22:01,708 Number of Imported URLs: 0
2017-11-22 16:22:01,708 Import warnings: []

// ... more unrelated output

8360 [ZAP-daemon] INFO org.zaproxy.zap.DaemonBootstrap  - ZAP is now listening on 0.0.0.0:39381
9502 [ZAP-Import-OpenAPI-1] WARN org.zaproxy.zap.extension.openapi.ExtensionOpenApi  - 
java.lang.NullPointerException
    at org.zaproxy.zap.extension.openapi.generators.PathGenerator.generateFullPath(PathGenerator.java:51)
    at org.zaproxy.zap.extension.openapi.converter.swagger.RequestModelConverter.generatePath(RequestModelConverter.java:63)
    at org.zaproxy.zap.extension.openapi.converter.swagger.RequestModelConverter.convert(RequestModelConverter.java:48)
    at org.zaproxy.zap.extension.openapi.converter.swagger.SwaggerConverter.convertToRequest(SwaggerConverter.java:75)
    at org.zaproxy.zap.extension.openapi.converter.swagger.SwaggerConverter.getRequestModels(SwaggerConverter.java:69)
    at org.zaproxy.zap.extension.openapi.ExtensionOpenApi$3.run(ExtensionOpenApi.java:212)
9691 [ZAP-ProxyThread-9] WARN org.zaproxy.zap.extension.api.API  - Bad request to API endpoint [/JSON/ascan/action/scan/] from [127.0.0.1]:
Provided parameter has illegal or unrecognized value (illegal_parameter) : url
    at org.zaproxy.zap.extension.ascan.ActiveScanAPI.scanURL(ActiveScanAPI.java:763)
    at org.zaproxy.zap.extension.ascan.ActiveScanAPI.handleApiAction(ActiveScanAPI.java:293)
    at org.zaproxy.zap.extension.api.API.handleApiRequest(API.java:431)
    at org.parosproxy.paros.core.proxy.ProxyThread.processHttp(ProxyThread.java:456)
    at org.parosproxy.paros.core.proxy.ProxyThread.run(ProxyThread.java:317)
    at java.lang.Thread.run(Thread.java:748)
14705 [ZAP-ProxyThread-10] WARN org.zaproxy.zap.extension.api.API  - Bad request to API endpoint [/JSON/ascan/view/status/] from [127.0.0.1]:
Does Not Exist (does_not_exist) : scanId
    at org.zaproxy.zap.extension.ascan.ActiveScanAPI.getActiveScan(ActiveScanAPI.java:654)
    at org.zaproxy.zap.extension.ascan.ActiveScanAPI.handleApiView(ActiveScanAPI.java:827)
    at org.zaproxy.zap.extension.api.API.handleApiRequest(API.java:467)
    at org.parosproxy.paros.core.proxy.ProxyThread.processHttp(ProxyThread.java:456)
    at org.parosproxy.paros.core.proxy.ProxyThread.run(ProxyThread.java:317)
    at java.lang.Thread.run(Thread.java:748)

ERROR: Job failed: exit status 1

```

See also [https://stackoverflow.com/questions/47450610/zap-api-scan-error-using-zap-api-scan-py](https://stackoverflow.com/q/47450610/839332) for more information about the issue and the command line arguments I used.

**Comment by psiinon** [comment]
I've opened a PR to fix the NPEs: https://github.com/zaproxy/zap-extensions/pull/1153 but I still dont think its parsing the defn right :/

**Comment by mattcopas** [comment]
I am also experiencing this issue - any updates? :)

**Comment by kingthorin** [comment]
@mattcopas @bravoman there's a major refactor of the OpenAPI addon currently in progress, also adds OpenAPI v3 support. Hopefully once that's merged and published it will address your concerns.

**Comment by mattcopas** [comment]
Thanks @kingthorin - any idea when this will be available?

**Comment by thc202** [comment]
@mattcopas the add-on was released today, could you give it a try?

**Comment by mattcopas** [comment]
@thc202 sorry for the late reply.

I won't be doing any security testing stuff for at least a week, but I'll give feedback when I do.

Meanwhile if anybody has tested this, please coment :)

**Comment by mattcopas** [comment]
@thc202 @kingthorin I'm available to test this now.

I'm doing this via Docker (currently using the `owasp/zap2docker-stable` image.

Is there a docker image available with the aforementioned addon?

Thanks for your help.

**Comment by thc202** [comment]
That image should do, the script updates the add-on. (The image `owasp/zap2docker-weekly` includes the latest version though.)

**Comment by mattcopas** [comment]
I can confirm it does not work with `stable` (as expected as this is 6 months out of date, as far as I know)

It does work with `weekly` :)

**Comment by mattcopas** [comment]
@thc202 do you know when this will be fixed in the `stable` version?

**Comment by thc202** [comment]
Odd, the script should be updating the add-on. Is there any error during start up when running with stable image?

The sable images are usually updated after/during the main release (2.9.0 in this case, still no fixed date).

**Comment by psiinon** [comment]
This could be a result of #5704 - maybe we should fix this for 2.9.0?

**Comment by thc202** [comment]
Closing, there's confirmation that this is fixed (https://github.com/zaproxy/zaproxy/issues/4072#issuecomment-566159097). Thank you!
